Synchronization settings and UTC time extension

raymond30031 opened the pull request:
Adds code and example for setting synchronization settings and UTC time

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1699,6 +1811,46 @@ def get_settings(arg):
+    """Parse command line synchronization-settings argument."""
+    # Parse each field from the argument
+    settings = arg.split(',')
+    settings = tuple([int(i) for i in settings]) 
```
It'd be better if you could protect this `int(i)` call with a try/except block (and a suitable error message).

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1699,6 +1811,46 @@ def get_settings(arg):
+    if valid:
+        return settings
+    else:
+        print "Invalide synchronization settings!"
```
`Invalide`->`Invalid`

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1699,6 +1811,46 @@ def get_settings(arg):
+    else:
+        valid = False
+
+    if valid:
```
I can see the interest of having a Boolean sometimes but here, I'd prefer if you could merge the two tests.

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1699,6 +1811,46 @@ def get_settings(arg):
+def get_UTCtime(arg):
+    # Parse each field from the argument
+    time_settings = arg.split(',')
+    time_settings = [int(i) for i in time_settings]
```
Same as above: protect the `int` call.

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1699,6 +1811,46 @@ def get_settings(arg):
+    if valid:
+        return time_settings
+    else:
+        print "Invalid UTCtime settings!"
```
Not sure the `!` is necessary, a full stop is probably good enough.

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1699,6 +1811,46 @@ def get_settings(arg):
+        valid = True
+    else:
+        valid = False
+    if valid:
```
Same as above: use the same test.

raymond30031 commented:
Hello Francis,

Sorry for the delay, I have finaly got around to address your comments.
Please review my latest commit.

Cheers,
JC

raymond30031 commented on nodes/mtdevice.py:
```diff
@@ -1254,6 +1258,85 @@ def usage():
+        day: day of the month, [1,31]
+        hour: hour of the day, [0,23]
+        min: minute of hour, [0,59]
+        sec: seconds of minute, minute [0,59]
```
I tested this. it carries over.

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1254,6 +1258,85 @@ def usage():
+            4 Interval Transition Measurement
+            8 SendLatest
+            9 ClockBiasEstimation
+            11 StartSampling
```
Ok, I was not clear, the two spaces I was talking about are between the letter/number and its description:
`11__StartSampling` instead of `11_StartSampling` (with space instead of underscore, of course).

fcolas commented on nodes/mtdevice.py:
```diff
@@ -577,7 +577,7 @@ def GetUTCTime(self):
         return (ns, year, month, day, hour, minute, second, flag)
 
-    def SetUTCTime(self, ns, year, month, day, hour, minute, second, flag):
+    def SetUTCTime(self, year, month, day, hour, minute, second, ns, flag):
```
No, don't change internal API, which is intended to follow the documentation and message. The change in order is for the command line interface only.

fcolas commented:
Thanks for your work. There are a few small comments but it should be easy to fix.

raymond30031 commented on nodes/mtdevice.py:
```diff
@@ -1318,14 +1316,14 @@ def usage():
+        hour: range [0,23]
+        min: range [0,59]
+        sec: range [0,59]
+        ns: range [0,1000000000]
```
I thought you wanted me to remove it because your previous comment said there are too many minutes

fcolas replied:
Ah, no, it's just that for the seconds, the description had the word minute twice instead of once.

raymond30031 commented:
Hello Francis,

I addressed your comments and extended the functionality for UTCtime settings.
We can now either set the current time or a specified time to be the UTCtime for the MTi.

Cheers,
JC

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1699,6 +1815,64 @@ def get_settings(arg):
+    # If argument is now, fill the time settings with the current time
+    # else fill the time settings with the specified time
+    if arg == "now":
+        timestamp2 = datetime.datetime.utcnow() # use datetime to get microsecond
```
That should be `timestamp` instead of `timestamp2`!

raymond30031 replied:
oops

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1254,6 +1259,87 @@ def usage():
+
+SetUTCTime settings:
+    There are two ways to set the UTCtime for the MTi.
+    Option #1: set MTi to the current UTC time based on local system time
```
It's really a great idea!
But you should say how to use this option (providing keyword 'now') in the usage description and not just in the example.

raymond30031 replied:
addressed

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1699,6 +1815,64 @@ def get_settings(arg):
+    # check synchronization settings
+    if settings[0] in (3, 4, 8, 9, 11) and \
+       settings[1] in (0, 1, 2, 4, 5, 6) and \
+       settings[2] in (1, 2, 3) and \
```
`settings[2]` can be also be `0`; that's actually the only way the clear the synchronization settings

raymond30031 replied:
I see what you mean now

raymond30031 replied:
added

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1370,6 +1459,16 @@ def main():
             actions.append('xkf-scenario')
+        elif o in ('-y', '--synchronization'):
+            sync_settings.append(get_synchronization_settings(a))
+            if sync_settings is None:
```
Argh. Doesn't work since your testing a list and not the last element: it should be something like:
```python
new_sync_setting = get_synchronization_settings(a)
if new_sync_setting is None:
    return 1
sync_settings.append(new_sync_setting)
```

raymond30031 replied:
good catch, addressed

fcolas commented on nodes/mtdevice.py:
```diff
@@ -1254,6 +1259,87 @@ def usage():
+    Note: The entire synchronization buffer is wiped every time a new one 
+          is set, so it is necessary to specify the settings of multiple 
+          lines at once.
+
```
You should also say how to clear the synchronization settings. Better yet, why not a keyword `clear` as in `SetUTCTime`? (which would send a single sync setting with polarity 0.)

raymond30031 replied:
Is the plan to create a function that clears all synchronization settings?

raymond30031 replied:
added

fcolas commented:
Dear JC,

Thanks again for your efforts. Please fix the typo for the time and the synchronization settings and I'll merge it in.

Best,

raymond30031 commented:
Hello Francis,

I have encountered a weird problem with my MTi G 710, and it does not happen to my MTi 300 or MTi 30.
Step 1:
Setting UTCtime to now and verifies using ./mtdevice -i to check the time is in fact the current time (my computer time)
command:
./mtdevice.py -c iu200,oq200fe,aa200fe,wr200fe,sw200 -u now

Step 2: 
roslaunch the xsens_driver launch file and check the time in the imu_data_str

The problem is that the time in the imu_data_str becomes sometime in 2013 and this is repeatable.
The flags is 240 in the time part of the message and the status word is 131 (which means no error)

Do you have any idea why this is happening?

raymond30031 commented:
Another odd thing I noticed today is that even though we assign iu and sw to be 200 and it does output at 200 Hz, but the output config from inspection shows that these two fields have the value of 65535 (maxed out for int?)

fcolas commented:
Many thanks for your work: it is merged in.

raymond30031 commented:
Glad I can contribute.
@fcolas
Did you get a chance to test -u now with a MTi-G-710 and see if the time gets reset upon startup?

fcolas commented:
Oops, I had missed this.
I wonder if it shouldn't be in a separate issue.
In the documentation, only the last 3 bits are meaningful with 0x04 stating whether it is fully valid or not. In your case it is 0 so it's not inconsistent. My hypothesis would be that it is searching for the GPS and outputting some initial value until it gets a fix. I agree it shouldn't be the case (with the `iu` option rather than `gu`), but I believe it's a firmware issue.
